### PR TITLE
dnsdist: Fix SNI on resumed sessions by acknowledging the name sent by the client

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Otherwise `SSL_get_servername()` only returns true when the session has been freshly established, and will return `nullptr` when it is resumed.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdistdist/libssl.cc [hunk 1]
```diff
@@ -685,6 +697,11 @@ std::unique_ptr<SSL_CTX, void(*)(SSL_CTX*)> libssl_init_server_context(const TLS
     SSL_CTX_sess_set_cache_size(ctx.get(), config.d_maxStoredSessions);
   }
 
+  /* we need to set this callback to acknowledge the server name sent by the client,
+     otherwise it will not stored in the session and will not be accessible when the
+     session is resumed, causing SSL_get_servername to return nullptr */
+  SSL_CTX_set_tlsext_servername_callback(ctx.get(), &libssl_server_name_callback);
```
Comment (omoerbeek) [inline pdns/dnsdistdist/libssl.cc:703]: extremely pedantic nit: redundant `&`

Comment (rgacogne) [reply pdns/dnsdistdist/libssl.cc:703]: I personally prefer it that way to make it more clear that we pass the address of the function, and not the result of a call to the function.

Comment (omoerbeek) [reply pdns/dnsdistdist/libssl.cc:703]: To my eyes it's clear it isn't a call since no arguments are passed. But fair enough.

**Comment by rgacogne** [comment]
Regression tests added.

**Pull request merged**